### PR TITLE
DeliverySpec.RetryAfter Experimental-Feature

### PR DESCRIPTION
**Pull request by travis-minke-sap**
This change id based on eventing [Issue #5811](https://github.com/knative/eventing/issues/5811) for which the initial implementation has been merged.

## Proposed Changes 

- Document proposed experimental-feature for supporting Retry-After headers.

Ready for review now that the initial Retry-After implementation has been merged!

### CODE REVIEW

**Comment by csantanapr** [comment]
This is ready for review @snneji if you want to take a look

**Comment by csantanapr** [comment]
@travis-minke-sap you don't need to put a hold for review, when something is in hold means we are going to ignore it and assume the PR submitter is still working on changes to be submitted and not ready yet

/unhold

File: docs/eventing/experimental-features.md [hunk 1]
```diff
@@ -103,6 +103,101 @@ affecting existing resources.
...
+largest of the normal backoff duration and the Retry-After header value will be
+used for the subsequent retry attempt. Specifying a "zero" value of "PT0S" will
+effectively disable **Retry-After** support.
+
+Prior to this experimental feature, Knative Eventing implementations have not
+supported **Retry-After** headers, and this is an attempt to provide a path
+for standardizing that support.  To begin the feature will be **opt-in**, but
+the final state will be **opt-out** as follows...
+
+<table>
```
Comment (csantanapr) [inline docs/eventing/experimental-features.md:138]: Can this table be in markdown?
Ref: https://github.github.com/gfm/#tables-extension-

Comment (travis-minke-sap) [reply docs/eventing/experimental-features.md]: definitely - have reworked it - thanks for reviewing!

File: docs/eventing/experimental-features.md [hunk 2]
```diff
@@ -103,6 +103,78 @@ affecting existing resources.
...
+`delivery` field.
+
+The `retryAfterMax` field will only take effect if the `delivery` spec is
+configured to perform retries, and will only pertain to retry attempts on
+**429** and **503** response codes. The field provides an override to prevent
+large **Retry-After** durations from impacting throughput, and must be specified
+using the [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601#Times) format. The
+largest of the normal backoff duration and the Retry-After header value will be
+used for the subsequent retry attempt. Specifying a "zero" value of "PT0S" will
+effectively disable **Retry-After** support.
```
Comment (snneji) [inline docs/eventing/experimental-features.md:131]:
```suggestion
used for the subsequent retry attempt. Specifying a "zero" value of `PT0S`
effectively disables **Retry-After** support.
```

File: docs/eventing/experimental-features.md [hunk 3]
```diff
@@ -103,6 +103,78 @@ affecting existing resources.
...
+
+Prior to this experimental feature, Knative Eventing implementations have not
+supported **Retry-After** headers, and this is an attempt to provide a path
+for standardizing that support.  To begin the feature will be **opt-in**, but
+the final state will be **opt-out** as follows...
```
Comment (snneji) [inline docs/eventing/experimental-features.md:136]:
```suggestion
for standardizing that support.  To begin, the feature is **opt-in**, but
the final state will be **opt-out** as follows:
```

File: docs/eventing/experimental-features.md [hunk 4]
```diff
@@ -103,6 +103,78 @@ affecting existing resources.
...
+
+| Feature Stage | Feature Flag | retryAfterMax Field Absent | retryAfterMax Field Present |
+| --- | --- | --- | --- |
+| Alpha / Beta | Disabled | Accepted by Webhook Validation & Retry-After headers NOT enforced | Rejected by WebHook Validation |
+| Alpha / Beta | Enabled | Accepted  by Webhook Validation & Retry-After headers NOT enforced | Accepted by Webhook Validation & Retry-After headers enforced if max override > 0 |
+| Stable / GA | n/a | Retry-After headers enforced without max override | Retry-After headers enforced if max override > 0 |
+
+The following example shows a Subscription that retries sending an event 3
```
Comment (snneji) [inline docs/eventing/experimental-features.md:144]:
```suggestion
The following example shows a Subscription that retries sending an event three
```

File: docs/eventing/experimental-features.md [hunk 5]
```diff
@@ -103,6 +103,78 @@ affecting existing resources.
...
+times, and respects **Retry-After** headers while imposing a max backoff of
```
Comment (snneji) [inline docs/eventing/experimental-features.md:145]:
```suggestion
times, and respects **Retry-After** headers while imposing a maximum backoff of
```

File: docs/eventing/experimental-features.md [hunk 6]
```diff
@@ -103,6 +103,78 @@ affecting existing resources.
...
+    backoffDelay: PT2S
+    backoffPolicy: linear
+    retry: 3
+    retryAfterMax: PT120S
+```
+
+!!! note
+    While the experimental feature flag will enforce all DeliverySpec usage of
+    the `retryAfterMax` field via Webhook validation, it does not guarantee all
+    implementations (Channels, Sources, etc.) will actually implement support
```
Comment (snneji) [inline docs/eventing/experimental-features.md:170]:
```suggestion
    While the experimental feature flag enforces all DeliverySpec usage of
    the `retryAfterMax` field through Webhook validation, it does not guarantee all
    implementations, such as Channels or Sources, actually implement support
```

File: docs/eventing/experimental-features.md [hunk 7]
```diff
@@ -103,6 +103,78 @@ affecting existing resources.
...
+    for the field.  The shared `HTTPMessageSender.SendWithRetries()` logic
+    has been enhanced to support this feature, and all implementations using it
+    to perform retries will automatically benefit.  Sandbox implementations NOT
```
Comment (snneji) [inline docs/eventing/experimental-features.md:173]:
```suggestion
    to perform retries will automatically benefit.  Sandbox implementations not
```

File: docs/eventing/experimental-features.md [hunk 8]
```diff
@@ -103,6 +103,78 @@ affecting existing resources.
...
+    based on this shared library (e.g. RabbitMQ, Google Pub/Sub, etc.) would
+    require additional development effort to respect/enforce the new
```
Comment (snneji) [inline docs/eventing/experimental-features.md:175]:
```suggestion
    based on this shared library, for example RabbitMQ or Google Pub/Sub, would
    require additional development effort to respect the
```

File: docs/eventing/experimental-features.md [hunk 9]
```diff
@@ -103,6 +103,78 @@ affecting existing resources.
...
+    `retryAfterMax`field.
```
Comment (snneji) [inline docs/eventing/experimental-features.md:176]:
```suggestion
    `retryAfterMax` field.
```

**Review by snneji — CHANGES_REQUESTED** [review]
A few style tweaks.

File: docs/eventing/experimental-features.md (hunk 7 above)
Comment (travis-minke-sap) [reply docs/eventing/experimental-features.md]: 👍 Will change!

File: docs/eventing/experimental-features.md [hunk 10]
```diff
@@ -103,6 +103,78 @@ affecting existing resources.
...
+When using the `delivery` spec to configure event delivery parameters, you can
+use the `retryAfterMax` field to specify how
+HTTP [Retry-After](https://datatracker.ietf.org/doc/html/rfc7231#section-7.1.3)
+headers are handled when calculating backoff times for retrying **429** and
+**503** responses. You can specify a `delivery` spec for Channels,
+Subscriptions, Brokers, Triggers, and any other resource spec that accepts the
+`delivery` field.
+
+The `retryAfterMax` field will only take effect if the `delivery` spec is
+configured to perform retries, and will only pertain to retry attempts on
```
[the remark this answers is not in the document]
Comment (travis-minke-sap) [reply docs/eventing/experimental-features.md]: 👍 Will change!

File: docs/eventing/experimental-features.md (hunk 2 above)
Comment (travis-minke-sap) [reply docs/eventing/experimental-features.md]: 👍 Will change!

File: docs/eventing/experimental-features.md (hunk 3 above)
Comment (travis-minke-sap) [reply docs/eventing/experimental-features.md]: 👍 Will change!

File: docs/eventing/experimental-features.md (hunk 4 above)
Comment (travis-minke-sap) [reply docs/eventing/experimental-features.md]: 👍 Will change!

File: docs/eventing/experimental-features.md (hunk 5 above)
Comment (travis-minke-sap) [reply docs/eventing/experimental-features.md]: 👍 Will change!

File: docs/eventing/experimental-features.md (hunk 6 above)
Comment (travis-minke-sap) [reply docs/eventing/experimental-features.md]: 👍 Will change!

File: docs/eventing/experimental-features.md (hunk 8 above)
Comment (travis-minke-sap) [reply docs/eventing/experimental-features.md]: 👍 Will change!

File: docs/eventing/experimental-features.md (hunk 9 above)
Comment (travis-minke-sap) [reply docs/eventing/experimental-features.md]: 👍 Will change!

**Comment by travis-minke-sap** [comment]
> A few style tweaks.

Thanks for the great review @snneji - much appreciated - I've made the suggested changes.

**Comment by snneji** [comment]
I added the hold label back as I see the issue https://github.com/knative/eventing/issues/5811 hasn't closed yet, although it seems that PR https://github.com/knative/eventing/pull/5813 related to that issue has been merged. You can unhold again if it's ready.

/lgtm
/approve

**Review by snneji — APPROVED** [review]

**Pull request merged**